taskcat/s3_sync.py

```diff
@@ -0,0 +1,157 @@
+        s3fl = self._get_s3_file_list(bucket, prefix)
+        self._sync(fl, s3fl, bucket, prefix, acl=acl)
+        return
+
+    def _hash_file(self, file_path, chunk_size=8 * 1024 * 1024):
+        # This is a bit funky because of the way multipart upload etags are done, they are a md5 of the md5's from each part with the number of parts appended
+        # credit to hyperknot https://github.com/aws/aws-cli/issues/2585#issue-226758933
+        md5s = []
+
+        with open(file_path, 'rb') as fp:
+            while True:
+                data = fp.read(chunk_size)
+                if not data:
+                    break
+                md5s.append(hashlib.md5(data))
+
+        if len(md5s) == 1:
+            return '"{}"'.format(md5s[0].hexdigest())
+
+        digests = b''.join(m.digest() for m in md5s)
+        digests_md5 = hashlib.md5(digests)
+        return '"{}-{}"'.format(digests_md5.hexdigest(), len(md5s))
+
+    def _get_local_file_list(self, path, include_checksums=True):
+        file_list = []
+        path = os.path.abspath(os.path.expanduser(path))
+        for root, dirs, files in os.walk(path):
+            relpath = os.path.relpath(root, path) + "/"
+            exclude_path = False
+            if relpath == './':
+                relpath = ""
+            for p in S3Sync._exclude_path_prefixes:
+                if relpath.startswith(p):
+                    exclude_path = True
+                    break
+            if not exclude_path:
+                for file in files:
+                    exclude = False
+                    for p in S3Sync._exclude_files:
+                        if fnmatch.fnmatch(file, p):
+                            exclude = True
+                            break
+                    if not exclude:
+                        full_path = root + "/" + file
+                        if include_checksums:
+                            checksum = self._hash_file(full_path)
+                        else:
+                            checksum = ""
+                        file_list.append([full_path, relpath + file, checksum])
+        return file_list
+
+    def _get_s3_file_list(self, bucket, prefix):
+        objects = []
+        more_objects = True
+        continuation_token = None
+        while more_objects:
+            if not continuation_token:
+                resp = self.s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix)
+            else:
+                resp = self.s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix, ContinuationToken=continuation_token)
+            if 'Contents' in resp:
+                objects += [[i['Key'][len(prefix):], i['ETag']] for i in resp['Contents']]
+            if 'NextContinuationToken' in resp.keys():
+                continuation_token = resp['NextContinuationToken']
+            else:
+                more_objects = False
+        return objects
+
+    def _sync(self, local_list, s3_list, bucket, prefix, acl, threads=16):
```

Comment:
Super nitpicky: 
I think this function should follow either of the following.
- The variables should be named such that it self-documents (not the case right now)
- In-line comments should be added at each step so someone else could follow the logic (also not the case right now, in my opinion)

Reply:
done